File: ina/code/VIII.java
/**
 * 
 */
package si.lj.uni.fri.lna.test.ina.labs;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.SortedMap;
import java.util.TreeMap;

import si.lj.uni.fri.lna.LNA;
import si.lj.uni.fri.lna.entity.graph.Clustering;
import si.lj.uni.fri.lna.entity.graph.Edge;
import si.lj.uni.fri.lna.entity.graph.Graph;
import si.lj.uni.fri.lna.entity.graph.ListGraph;
import si.lj.uni.fri.lna.entity.graph.Metric;
import si.lj.uni.fri.lna.entity.graph.Node;
import si.lj.uni.fri.lna.entity.graph.Ranking;
import si.lj.uni.fri.lna.io.graph.Readers;
import si.lj.uni.fri.lna.logic.graph.Clusters;
import si.lj.uni.fri.lna.logic.graph.Communities;
import si.lj.uni.fri.lna.logic.graph.Components;
import si.lj.uni.fri.lna.logic.graph.Edges;
import si.lj.uni.fri.lna.logic.graph.Generators;
import si.lj.uni.fri.lna.logic.graph.Nodes;
import si.lj.uni.fri.lna.logic.utility.Comparison;
import si.lj.uni.fri.lna.logic.utility.Estimation;
import si.lj.uni.fri.lna.logic.utility.Generation;
import si.lj.uni.fri.lna.logic.utility.Implementation;
import si.lj.uni.fri.lna.logic.utility.Information;
import si.lj.uni.fri.lna.utility.Complexity;
import si.lj.uni.fri.lna.utility.Uncertainty;

/**
 * @author Lovro Subelj
 *
 */
public class VIII {
	
	public static void main(String[] args) throws Exception {
		LNA.init();
		
		/* implementations of community detection algorithms */
		List<Implementation<Clustering>> implementations = new ArrayList<Implementation<Clustering>>();
		implementations.add(new Implementation<Clustering>("Truth") {
			public Clustering run(Graph graph) throws Exception { return Clusters.clustering(graph); }});
		implementations.add(new Implementation<Clustering>("Labels") { 
			public Clustering run(Graph graph) throws Exception { return labels(graph); }});
		implementations.add(new Implementation<Clustering>("Louvain") {  
			public Clustering run(Graph graph) throws Exception { return Communities.louvain(graph); }});
		implementations.add(new Implementation<Clustering>("Infomap") { 
			public Clustering run(Graph graph) throws Exception { return Communities.infomap(graph); }});
		implementations.add(new Implementation<Clustering>("Graclus") { 
			public Clustering run(Graph graph) throws Exception { return Communities.graclus(graph); }});
		implementations.add(new Implementation<Clustering>("Between") { 
			public Clustering run(Graph graph) throws Exception { return betweenness(graph); }});
		
		String folder = "/Users/lovre/Documents/office/coding/repositories/www/ina/";
		
		/* small social networks with sociological partitioning */
		List<Graph> graphs = new ArrayList<Graph>();
		graphs.add(Readers.pajek(folder + "karate_club.net")); 
		graphs.add(Readers.pajek(folder + "american_football.net"));
		graphs.add(Readers.pajek(folder + "dolphins.net")); 
		graphs.add(Readers.pajek(folder + "southern_women.net"));  
		
		/* community detection in small social networks */
		for (Graph graph: graphs)
			Comparison.metrics(info(graph), implementations, 100);
		
		/* larger networks with node metadata */ graphs.clear();
		graphs.add(Readers.pajek(folder + "cdn_java.net"));
		graphs.add(Readers.pajek(folder + "dormitory.net"));
		graphs.add(Readers.pajek(folder + "wikileaks.net"));
		graphs.add(Readers.pajek(folder + "youtube.net"));
		
		/* community detection in larger networks */
		for (Graph graph: graphs)
			Comparison.metrics(info(graph), implementations.subList(0, 4), 10);
		
		/* community detection in GN synthetic graphs */
		Comparison.metrics(new Generation("GN", 11) { 
			public String name() { return "Mixing"; } public String name(int index) { return String.format("%12.2f | ", 0.05 * index); } 
			public Graph graph(int index) throws Exception { return Generators.array(GN(0.05 * index)); }
		}, implementations.subList(0, 4), Metric.NMI, 25);
		
		/* community detection in LFR synthetic graphs */
		Comparison.metrics(new Generation("LFR", 9) { 
			public String name() { return "Mixing"; } public String name(int index) { return String.format("%12.2f | ", 0.1 * index); } 
			public Graph graph(int index) throws Exception { return Generators.array(Generators.LFR(1000, 0.1 * index)); }
		}, implementations.subList(0, 4), Metric.NMI, 10);
		
		/* community detection in random graphs */
		Comparison.metrics(new Generation("ER", 9) { 
			public String name() { return "Degree"; } public String name(int index) { return String.format("%12.1f | ", 8.0 + 4.0 * index); } 
			public Graph graph(int index) throws Exception { return Generators.array(Generators.random(1000, 8.0 + 4.0 * index)); }}, 
			implementations.subList(0, 4), Metric.NMI, 10);
		
		LNA.exit();
	}
	
	static Graph info(Graph graph) throws Exception {
		graph = Generators.array(graph);
		
		Information.println("\n         ... | GRAPH"); 
		Information.print(graph, true); Complexity.poc();
		
		return graph;
	}

	static Graph GN(double random) {
		ListGraph graph = Generators.isolate("GN", 128);
		for (Node node: graph.getNodes())
			node.setCluster(node.getIndex() / 32);

		for (int i = 0; i < 128; i++)
			for (int j = i + 1; j < 128; j++)
				if (Uncertainty.binary(i / 32 == j / 32? 16.0 * (1.0 - random) / 31.0: random / 6.0))
					graph.addEdge(i, j);
				
		return graph;
	}
	
	static Clustering labels(Graph graph) {
		Clustering clustering = new Clustering(graph);
		
		Ranking ranking = new Ranking(graph);
		List<Integer> nodes = Nodes.nodes(graph);
		for (int node: nodes) {
			clustering.put(node, node);
			ranking.put(node, 1.0);
		}

		boolean converged = false;
		while (!converged) {
			Collections.shuffle(nodes);

			converged = true;
			for (int node: nodes) 
				if (graph.getDegree(node) > 0) {
					Map<Integer, Double> clusters = new HashMap<Integer, Double>();
					for (int neighbor: graph.getNeighbors(node))
						if (clusters.containsKey(clustering.get(neighbor)))
							clusters.put(clustering.get(neighbor), clusters.get(clustering.get(neighbor)) + ranking.get(neighbor));
						else
							clusters.put(clustering.get(neighbor), ranking.get(neighbor));

					double maximum = Estimation.maximum(clusters.values());
					if (!clusters.containsKey(clustering.get(node)) || clusters.get(clustering.get(node)) < maximum) {
						List<Integer> labels = new ArrayList<Integer>();
						for (int label: clusters.keySet())
							if (clusters.get(label) == maximum)
								labels.add(label);
						
						clustering.put(node, Uncertainty.random(labels));

						converged = false;
					}
				}
		}

		return clustering;
	}

	static Clustering betweenness(Graph graph) {
		Clustering clustering = new Clustering(graph);

		Graph copy = Generators.bag(graph).setUndirected();
		
		double Q = -Double.MAX_VALUE;
		while (copy.getM() > 0) {
			Map<Edge, Double> betweennesses = Edges.betweennesses(copy);
			double maximum = Estimation.maximum(betweennesses.values());
			
			List<Edge> edges = new ArrayList<Edge>();
			for (Edge edge: betweennesses.keySet())
				if (betweennesses.get(edge) == maximum)
					edges.add(edge);

			copy.removeEdge(Uncertainty.random(edges));

			SortedMap<Integer, Set<Integer>> clusters = new TreeMap<Integer, Set<Integer>>();
			for (Set<Integer> component: Components.WCCs(copy))
				clusters.put(clusters.size() + 1, component);

			double q = Clusters.Q(graph, clusters);
			if (q > Q) {
				clustering = Clusters.clustering(graph, clusters);
				Q = q;
			}
		}

		return clustering;
	}

}

File: ina/code/viii.py
import random

import networkx as nx
from cdlib.classes import *
from cdlib import viz, algorithms

from matplotlib import pyplot as plt

PATH = "/Users/lovre/Downloads"

def read(name, path = PATH):
  G = nx.MultiGraph(name = name)
  with open(path + "/" + name + ".net", 'r') as file:
    file.readline()

    for line in file:
      if line.startswith("*"):
        break
      else:
        node = line.split("\"")
        G.add_node(int(node[0]) - 1, label = node[1], cluster = int(node[2]) if len(node) > 2 and len(node[2].strip()) > 0 else 0)

    for line in file:
      i, j = (int(x) - 1 for x in line.split()[:2])
      G.add_edge(i, j)
      
  return G
  
def info(G):
  print("{:>12s} | '{:s}'".format('Graph', G.name))

  n = G.number_of_nodes()
  m = G.number_of_edges()
  
  print("{:>12s} | {:,d} ({:,d})".format('Nodes', n, nx.number_of_isolates(G)))
  print("{:>12s} | {:,d} ({:,d})".format('Edges', m, nx.number_of_selfloops(G)))
  print("{:>12s} | {:.2f} ({:,d})".format('Degree', 2 * m / n, max([k for _, k in G.degree()])))
  print("{:>12s} | {:.2e}".format('Density', 2 * m / n / (n - 1)))
  print()
  
def girvan_newman(mu = 0.5):
  G = nx.Graph(name = "girvan_newman")
  for i in range(128):
    G.add_node(i, cluster = i // 32 + 1)
  for i in range(128):
    for j in range(i + 1, 128):
      if G.nodes[i]['cluster'] == G.nodes[j]['cluster']:
        if random.random() < 16 * (1 - mu) / 31:
          G.add_edge(i, j)
      else:
        if random.random() < mu / 6:
          G.add_edge(i, j)
  return G
  
def compare(G, algs, P, num = 1):
  print("{:>12s} | {:5s} {:^6s}  {:^5s}  {:^5s}  {:^5s}".format('Algorithm', 'Count', 'Q', 'NMI', 'ARI', 'VI'))
  for alg in sorted(algs.keys(), reverse = True):
    c, Q, NMI, ARI, VI = 0, 0, 0, 0, 0
    for _ in range(num):
      C = algs[alg](G)
      c += len(C.communities) / num
      Q += C.newman_girvan_modularity().score / num
      NMI += C.normalized_mutual_information(P).score / num
      ARI += C.adjusted_rand_index(P).score / num
      VI += C.variation_of_information(P).score / num
    print("{:>12s} | {:>5.0f} {:6.3f}  {:5.3f}  {:5.3f}  {:5.3f}".format('\'' + alg + '\'', c, Q, NMI, ARI, VI))
  print()

def benchmarks(algs, mus = [0.05 * i for i in range(1, 13)], num = 1):
  info(girvan_newman())
  print("{:>12s} |".format('Mixing'), end = "")
  for mu in mus:
    print(" {:^5.3f}".format(mu), end = "")
  print()
  for alg in sorted(algs.keys(), reverse = True):
    print("{:>12s} |".format('\'' + alg + '\''), end = "")
    for mu in mus:
      NMI = 0
      for _ in range(num):
        G = girvan_newman(mu)
        C = algs[alg](G)
        NMI += C.normalized_mutual_information(partition(G)).score / num
      print(" {:5.3f}".format(NMI), end = "")
    print()
  print()

def randoms(algs, n = 2500, ks = list(range(1, 24, 3)), num = 1):
  G = nx.gnm_random_graph(n, n * ks[len(ks) // 2] // 2)
  G.name = "erdos_renyi"
  info(G)
  print("{:>12s} |".format('Degree'), end = "")
  for k in ks:
    print(" {:^5d}".format(k), end = "")
  print()
  for alg in sorted(algs.keys(), reverse = True):
    print("{:>12s} |".format('\'' + alg + '\''), end = "")
    for k in ks:
      VI = 0
      for _ in range(num):
        G = nx.gnm_random_graph(n, n * k // 2)
        C = algs[alg](G)
        VI += C.variation_of_information(components(G)).score / num
      print(" {:5.3f}".format(VI), end = "")
    print()
  print()
  
def partition(G):
  P = {}
  for node in G.nodes(data = True):
    if node[1]['cluster'] not in P:
      P[node[1]['cluster']] = []
    P[node[1]['cluster']].append(node[0])
  return NodeClustering(list(P.values()), G, 'Truth')
  
def components(G):
  return NodeClustering(list(nx.connected_components(G)), G)
  
algs = {"Truth": lambda G: partition(G), "Louvain": algorithms.louvain, "Infomap": algorithms.infomap, "LPA": algorithms.label_propagation, "EB": lambda G: algorithms.girvan_newman(G, level = 1), "SBM": algorithms.sbm_dl_nested}

for name in ["karate_club", "southern_women", "dolphins", "american_football"]:
  G = read(name)

  info(G)

  P = partition(G)

  compare(G, algs, P, num = 100)

  # viz.plot_network_clusters(G, algs["Truth"](G), nx.spring_layout(G))
  # plt.show()

algs.pop("EB")

benchmarks(algs, num = 100)

algs.pop("SBM")

for name in ["cdn_java", "dormitory", "wikileaks", "youtube"]:
  G = read(name)

  info(G)

  P = partition(G)

  compare(G, algs, P, num = 10)

algs.pop("Truth")

randoms(algs, num = 10)
